catalyst-gateway/bin/src/service/api/cardano/rbac/registrations_get/chain_info.rs

```diff
@@ -0,0 +1,134 @@
+        catalyst_id: &IdUri,
+    ) -> anyhow::Result<Option<Self>> {
+        let (persistent_registrations, volatile_registrations) = try_join(
+            indexed_registrations(persistent_session, catalyst_id),
+            indexed_registrations(volatile_session, catalyst_id),
+        )
+        .await?;
+        let network = Settings::cardano_network();
+
+        let (chain, last_persistent_slot) =
+            apply_registrations(network, None, persistent_registrations).await?;
+        let last_persistent_txn = chain.as_ref().map(RegistrationChain::current_tx_id_hash);
+
+        let (chain, _) = apply_registrations(network, chain, volatile_registrations).await?;
+        let mut last_volatile_txn = chain.as_ref().map(RegistrationChain::current_tx_id_hash);
+        if last_persistent_txn == last_volatile_txn {
+            last_volatile_txn = None;
+        }
+
+        Ok(chain.map(|chain| {
+            Self {
+                chain,
+                last_persistent_txn,
+                last_volatile_txn,
+                last_persistent_slot,
+            }
+        }))
+    }
+}
+
+/// Returns a sorted list of all registrations for the given Catalyst ID from the
+/// database.
+async fn indexed_registrations(
+    session: &CassandraSession, catalyst_id: &IdUri,
+) -> anyhow::Result<Vec<Query>> {
+    let mut result: Vec<_> = Query::execute(session, QueryParams {
+        catalyst_id: catalyst_id.clone().into(),
+    })
+    .and_then(|r| r.try_collect().map_err(Into::into))
+    .await?;
+
+    result.sort_by_key(|r| r.slot_no);
+    Ok(result)
+}
+
+/// Applies the given list of indexed registrations to the chain. If the given chain is
+/// `None` - new root will be possibly created.
+async fn apply_registrations(
+    network: Network, mut chain: Option<RegistrationChain>, indexed_registrations: Vec<Query>,
+) -> anyhow::Result<(Option<RegistrationChain>, Slot)> {
+    let mut last_slot = 0.into();
+
+    for reg in indexed_registrations {
+        // We perform validation during indexing, so this normally should never fail.
+        let cip509 = registration(network, reg.slot_no.into(), reg.txn_index.into())
+            .await
+            .with_context(|| {
+                format!(
+                    "Failed to get Cip509 registration from {:?} slot and {:?} txn index",
+                    reg.slot_no, reg.txn_index
+                )
+            })?;
+
+        match chain {
+            None => {
+                if let Ok(root) = RegistrationChain::new(cip509) {
+                    chain = Some(root);
+                    last_slot = reg.slot_no.into();
+                }
+            },
+            Some(ref current) => {
+                // This isn't a hard error because while the individual registration can
+                // be valid it can be invalid in the context of the whole registration
+                // chain.
+                if let Ok(new) = current.update(cip509) {
```

Comment:
If the registration chain is

`Chain root role 0 -> Reg 2 role 0 -> Reg 3 role 0 -> Reg 4 Role 1`

The chain wont be constructed properly, because

The `.update` function is checking
```
if cip509.catalyst_id().is_some() {
            bail!("Catalyst id should be present only for chain root registration.");
        }
```
where `catalyst_id` is set for Role 0



----------------------------------------

The scenario that this will work is 
`chain root reg 1 role 0 -> reg 2 none 0 role`

Reply:
(Hopefully) fixed. 🙃

Reply:
https://github.com/input-output-hk/catalyst-libs/pull/268